### PR TITLE
Correctly generate record labels with foreign keys

### PR DESCRIPTION
**Pull request by aschempp**
This fixes a long standing issue that record labels are not correctly generated from foreignKey. I noticed this because of a strange behaviour: I have a custom `label_callback`, but the column view did not show the value my callback returned. This was because the value I returned was used to look up the foreignKey! And for one of 10 records, it actually found something in the foreign tabel (for no good reason) and showed that instead.

### CODE REVIEW

**Review by Toflar — DISMISSED** [review]
😂

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 1]
```diff
@@ -1707,7 +1707,18 @@ public function generateRecordLabel(array $row, string $table = null, bool $prot
 					->limit(1)
 					->execute($row[$strKey]);
 
-				$args[$k] = $objRef->numRows ? $objRef->$strField : '';
+				$args[$k] = $objRef->numRows ? $objRef->$strField : '-';
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]: Why did you change the default value?

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 2]
```diff
@@ -1726,7 +1737,7 @@ public function generateRecordLabel(array $row, string $table = null, bool $prot
 			}
 			elseif (($GLOBALS['TL_DCA'][$table]['fields'][$v]['eval']['isBoolean'] ?? null) || (($GLOBALS['TL_DCA'][$table]['fields'][$v]['inputType'] ?? null) == 'checkbox' && !($GLOBALS['TL_DCA'][$table]['fields'][$v]['eval']['multiple'] ?? null)))
 			{
-				$args[$k] = $row[$v] ? $GLOBALS['TL_LANG']['MSC']['yes'] : $GLOBALS['TL_LANG']['MSC']['no'];
+				$args[$k] = ($row[$v] ?? null) ? $GLOBALS['TL_LANG']['MSC']['yes'] : $GLOBALS['TL_LANG']['MSC']['no'];
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/DataContainer.php:1740]: This seems unrelated.

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 3]
```diff
@@ -1755,10 +1766,12 @@ public function generateRecordLabel(array $row, string $table = null, bool $prot
 				{
 					$args[$k] = $row[$v];
 				}
+
+				$args[$k] = (string) $args[$k] !== '' ? $args[$k] : '-';
 			}
 			else
 			{
-				$args[$k] = null;
+				$args[$k] = '-';
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]: This will have side-effects. Please revert.

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/DataContainer.php:1740]: it is the last remaining place in this method that does not check if `isset($row[$v])`. All the other places have been fixed before, so this should be fixed as well.

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 3 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: what side effects are you expecting? Since the same default value is applied above, e.g. if a column is `rgxp=date` but the value is empty?

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: Because this is the same default value as all the other if-conditions in the method will return. It is previously done in the `listView`, but should be unified across all views.

File: core-bundle/src/Resources/contao/drivers/DC_Table.php [hunk 4]
```diff
@@ -5106,30 +5106,6 @@ protected function listView()
...
-								}
-							}
-
-							$value = $arg ?: '-';
-						}
-						else
-						{
-							$value = (string) $arg !== '' ? $arg : '-';
-						}
-
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/drivers/DC_Table.php]: Removing this causes an error on my system:

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 3 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: This is the status quo:



If the first value is `null`, we e.g. replace the value with an icon.

But now the first value is no longer empty:



This will have side-effects for sure!

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: This needs to be reverted, too (see below).

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 3 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: why would we not replace the icon if the value is not null? `tl_member:addIcon` always adds the icon at least.
Is this status quo the 4th callback argument?

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/DataContainer.php]: I have resolved the dash in 5265c250436d5b54e35fd4166a2e7821a1607742 since I noticed there are other places that rely on an empty value (e.g. the Picker widget).

File: core-bundle/src/Resources/contao/drivers/DC_Table.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/drivers/DC_Table.php]: Fixed in 5265c250436d5b54e35fd4166a2e7821a1607742

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.
